ext/mbstring/tests/gb18030_2022_encoding.phpt

```diff
@@ -0,0 +1,311 @@
+ "\x00\x08\x20\xF6" => "\xB5\x30\xE4\x30",
+ "\x00\x08\xAD\x05" => "\xB7\x39\x9F\x35",
+ "\x00\x08\xEA\x7E" => "\xB9\x31\xDD\x32",
+ "\x00\x08\xF0\xB8" => "\xB9\x32\xFE\x36",
+ "\x00\x09\x14\x07" => "\xBA\x30\x96\x35",
+ "\x00\x09\x41\xDD" => "\xBA\x39\xBD\x39",
+ "\x00\x09\x42\xEF" => "\xBA\x39\xD9\x33",
+ "\x00\x09\xBA\x2B" => "\xBD\x33\xF5\x37",
+ "\x00\x0A\x26\x00" => "\xBF\x35\xEA\x32",
+ "\x00\x0A\x36\xE9" => "\xBF\x39\xA3\x31",
+ "\x00\x0A\x7A\x20" => "\xC1\x32\xF5\x38",
+ "\x00\x0A\x9C\x93" => "\xC1\x39\xF5\x37",
+ "\x00\x0A\xC0\xD7" => "\xC2\x37\xA6\x31",
+ "\x00\x0A\xD8\x77" => "\xC3\x32\x8C\x39",
+ "\x00\x0B\x1A\x9B" => "\xC4\x35\xC4\x31",
+ "\x00\x0B\x4F\x27" => "\xC5\x36\x9B\x33",
+ "\x00\x0B\x72\x6D" => "\xC6\x33\xB0\x33",
+ "\x00\x0B\xEE\x23" => "\xC8\x38\xC1\x33",
+ "\x00\x0B\xF0\xDF" => "\xC8\x39\x89\x33",
+ "\x00\x0C\x0B\xE1" => "\xC9\x34\xC6\x37",
+ "\x00\x0C\x4C\x98" => "\xCA\x37\xD9\x34",
+ "\x00\x0C\x5F\x41" => "\xCB\x31\xBF\x31",
+ "\x00\x0C\x63\xE4" => "\xCB\x32\xB7\x38",
+ "\x00\x0C\x70\x0A" => "\xCB\x34\xF2\x38",
+ "\x00\x0C\xAD\x6A" => "\xCC\x37\xB0\x30",
+ "\x00\x0C\xCC\x03" => "\xCD\x33\xCB\x33",
+ "\x00\x0C\xD5\x4C" => "\xCD\x35\xBD\x30",
+ "\x00\x0C\xE6\x70" => "\xCD\x38\xF9\x38",
+ "\x00\x0D\x1B\x6A" => "\xCE\x39\xDC\x30",
+ "\x00\x0D\x55\xEE" => "\xD0\x31\xCE\x30",
+ "\x00\x0D\xBB\xB1" => "\xD2\x32\xA5\x31",
+ "\x00\x0D\xC0\x4F" => "\xD2\x33\x9D\x33",
+ "\x00\x0D\xFA\x84" => "\xD3\x35\x87\x34",
+ "\x00\x0E\x16\x71" => "\xD4\x30\xDC\x33",
+ "\x00\x0E\x1E\x03" => "\xD4\x32\xA2\x31",
+ "\x00\x0E\x20\xE8" => "\xD4\x32\xEC\x32",
+ "\x00\x0E\x39\x6A" => "\xD4\x37\xE9\x36",
+ "\x00\x0E\x6A\x95" => "\xD5\x37\xE8\x33",
+ "\x00\x0E\x7E\xCD" => "\xD6\x31\xF5\x39",
+ "\x00\x0E\x80\x69" => "\xD6\x32\xA1\x31",
+ "\x00\x0E\x9A\x7F" => "\xD6\x37\xC6\x39",
+ "\x00\x0E\xEE\x12" => "\xD8\x34\xC4\x34",
+ "\x00\x0E\xFC\xA1" => "\xD8\x37\xBF\x31",
+ "\x00\x0F\x29\xB0" => "\xD9\x36\xD2\x36",
+ "\x00\x0F\x2A\x12" => "\xD9\x36\xDC\x34",
+ "\x00\x0F\x6C\x8C" => "\xDB\x30\x9E\x32",
+ "\x00\x0F\xAF\x04" => "\xDC\x33\xDD\x38",
+ "\x00\x0F\xBE\x65" => "\xDC\x36\xED\x35",
+ "\x00\x0F\xE5\x88" => "\xDD\x34\xE7\x34",
+ "\x00\x0F\xE7\xB1" => "\xDD\x35\xA0\x37",
+ "\x00\x0F\xF4\x27" => "\xDD\x37\xE3\x37"];
+testAllValidChars($sampleSMP, 'UTF-32BE', 'GB18030-2022', false);
+
+function readGB18030_2022_ConversionTable($path, &$from, &$to, $utf32 = false) {
+    $from = [];
+    $to   = [];
+
+    $fp = fopen($path, 'r+');
+    while ($line = fgets($fp, 256)) {
+        if ($line[0] == '#')
+            continue;
+        if (sscanf($line, "%x\t%x", $codepoint, $char) == 2) {
+            $codepoint = $utf32 ? pack('N', $codepoint) : pack('n', $codepoint);
+            if ($char == PHP_INT_MAX) {
+                // We may be on a 32-bit machine and testing a text encoding with 4-byte codes
+                // (which can't be represented in a PHP integer)
+                $char = "";
+                for ($i = 2; $i < strlen($line); $i += 2) {
```

Comment:
I bet the 32-bit failure is caused by this line, the offset `2` doesn't make sense, I think it should be `strpos($line, "\t") + 1`
Will check tonight.

Also this test would need `ctype` extension as dependency on 32-bit platforms.

Reply:
That totally makes sense.

Reply:
See https://github.com/php/php-src/pull/13069